# ML-08 — Capstone Modeling Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/AxelYoel/FlyRank-AI-Internship---Axel-Yoel-Chandra/blob/main/work/notebooks/w05_model.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Method choice and why

The question is "which pages should be reviewed first?", so this is a ranking task.
I predict each page's lost_clicks and rank by the prediction. I compared:
- the W04 rule (capture_ratio) as the baseline
- Linear Regression (readable, tried first) and Random Forest (a stronger candidate)
- two no-model reference rows: rank by February impressions only, and rank by
  February's lost_clicks (carry last month forward)
Complexity has to earn its place, so the simple references are in the table too.

In [21]:
%pip -q install duckdb
import duckdb, os
from google.colab import userdata

hf_token = userdata.get("HF_token")  # exact name you set in Colab Secrets
os.environ["HF_token"] = hf_token

con = duckdb.connect()
con.execute(f"CREATE OR REPLACE SECRET hf (TYPE huggingface, TOKEN '{hf_token}')")
rel = "hf://datasets/FlyRank/internship-warehouse"

In [22]:
con.sql(f"DESCRIBE SELECT * FROM read_parquet('{rel}/dim_content.parquet')").df()

,column_name,column_type,null,key,default,extra
0,client_hash_id,VARCHAR,YES,None,None,None
1,content_hash_id,VARCHAR,YES,None,None,None
2,keyword_hash_id,VARCHAR,YES,None,None,None
3,url_hash_id,VARCHAR,YES,None,None,None
4,keyword_char_count,BIGINT,YES,None,None,None
5,keyword_token_count,BIGINT,YES,None,None,None
6,url_char_count,BIGINT,YES,None,None,None
7,content_created_date,DATE,YES,None,None,None
8,content_updated_date,DATE,YES,None,None,None
9,content_type,VARCHAR,YES,None,None,None


In [23]:
con.sql(f"DESCRIBE SELECT * FROM read_parquet('{rel}/dim_clients.parquet')").df()

,column_name,column_type,null,key,default,extra
0,client_hash_id,VARCHAR,YES,None,None,None
1,is_active,BOOLEAN,YES,None,None,None
2,has_gsc_access,BOOLEAN,YES,None,None,None
3,has_ga4_access,BOOLEAN,YES,None,None,None
4,access_profile,VARCHAR,YES,None,None,None
5,client_created_date,DATE,YES,None,None,None
6,client_updated_date,DATE,YES,None,None,None
7,gsc_data_start,DATE,YES,None,None,None
8,ga4_data_start,DATE,YES,None,None,None


In [24]:
scored = con.sql(f"""
    WITH daily AS (
        SELECT content_hash_id, gsc_avg_position, gsc_impressions, gsc_clicks
        FROM read_parquet('{rel}/fact_content_daily_performance/month=2026-03/data_0.parquet')
        WHERE gsc_data_available = TRUE AND gsc_avg_position > 0 AND gsc_impressions > 0
    ),
    rollup AS (
        SELECT content_hash_id, SUM(gsc_impressions) AS impressions, SUM(gsc_clicks) AS clicks
        FROM daily
        GROUP BY content_hash_id
        HAVING SUM(gsc_impressions) >= 50
    )
    SELECT r.content_hash_id, r.impressions, r.clicks,
           c.client_hash_id, c.content_type, c.search_volume
    FROM rollup r
    JOIN read_parquet('{rel}/dim_content.parquet') c
        ON r.content_hash_id = c.content_hash_id
    WHERE c.content_type = 'keyword article' AND c.search_volume >= 20
""").df()

print(f"Scored population: {len(scored)} rows")

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Scored population: 38726 rows


In [25]:
n_clients = scored['client_hash_id'].nunique()
print(f"Distinct clients in scored population: {n_clients}")

Distinct clients in scored population: 43


In [26]:
pages_per_client = scored.groupby('client_hash_id').size()
print(pages_per_client.describe())
print()
print(pages_per_client.quantile([0.1, 0.25, 0.5, 0.75, 0.9, 0.99]))

count       43.000000
mean       900.604651
std       2146.087316
min          1.000000
25%          6.500000
50%         64.000000
75%        703.500000
max      11127.000000
dtype: float64

0.10       2.20
0.25       6.50
0.50      64.00
0.75     703.50
0.90    2795.20
0.99    9462.12
dtype: float64


In [27]:
pages_per_client.sort_values(ascending=False).head(10)

,0
client_hash_id,
client_73cda7b4e4f265ea,11127
client_62f4a7e64f5e0096,7163
client_e547b89c05043229,4415
client_08a6a72ff48e62c0,4118
client_fef1a8f436438636,3156
client_2094c6eb080311d5,1352
client_e5c2aa26a8598242,1054
client_23a62021009f63c4,842
client_a80fca3f171ed1de,814


## 2. Split design

Grouped k-fold by client (k=5 and k=10), 43 clients, 38,726 rows in the scored slice.
A random split would let the same client's pages sit in both train and test.
The data is very uneven: the top 5 clients hold about 77% of the rows, so
each giant client ends up alone in a test fold. Folds 0 and 1 are the same in
k=5 and k=10, so the two k values are not independent evidence. Seeds fixed at 42.
March is the target month, February is used as a feature, June stays sealed.

In [28]:
import numpy as np
import pandas as pd
from sklearn.model_selection import GroupKFold
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score

In [29]:
model_data = con.sql(f"""
    WITH daily AS (
        SELECT content_hash_id, gsc_avg_position, gsc_impressions, gsc_clicks
        FROM read_parquet('{rel}/fact_content_daily_performance/month=2026-03/data_0.parquet')
        WHERE gsc_data_available = TRUE AND gsc_avg_position > 0 AND gsc_impressions > 0
    ),
    rollup AS (
        SELECT
            content_hash_id,
            SUM(gsc_impressions) AS impressions,
            SUM(gsc_clicks) AS clicks,
            SUM(gsc_avg_position * gsc_impressions) / SUM(gsc_impressions) AS weighted_avg_position
        FROM daily
        GROUP BY content_hash_id
        HAVING SUM(gsc_impressions) >= 50
    ),
    tiered AS (
        SELECT *,
            CASE
                WHEN weighted_avg_position <= 3 THEN 'tier_1_1-3'
                WHEN weighted_avg_position <= 10 THEN 'tier_2_4-10'
                WHEN weighted_avg_position <= 20 THEN 'tier_3_11-20'
                ELSE 'tier_4_21plus'
            END AS position_tier
        FROM rollup
    ),
    benchmark(position_tier, expected_ctr) AS (
        VALUES
            ('tier_1_1-3', 0.002811), ('tier_2_4-10', 0.002353),
            ('tier_3_11-20', 0.002226), ('tier_4_21plus', 0.000748)
    ),
    labeled AS (
        SELECT
            t.content_hash_id, t.impressions, t.clicks, t.weighted_avg_position,
            b.expected_ctr,
            (b.expected_ctr * t.impressions) - t.clicks AS lost_clicks
        FROM tiered t
        JOIN benchmark b USING (position_tier)
    )
    SELECT
        l.content_hash_id, c.client_hash_id,
        l.lost_clicks, l.impressions, l.clicks, l.weighted_avg_position, l.expected_ctr,
        c.search_volume, c.competition_level, c.content_type,
        CASE WHEN c.content_updated_date <= DATE '2026-03-31'
             THEN DATE_DIFF('day', c.content_updated_date, DATE '2026-03-31')
             ELSE NULL END AS days_since_update_capped,
        c.word_count
    FROM labeled l
    JOIN read_parquet('{rel}/dim_content.parquet') c
        ON l.content_hash_id = c.content_hash_id
    WHERE c.content_type = 'keyword article' AND c.search_volume >= 20
""").df()

print(f"Rows: {len(model_data)}, distinct clients: {model_data['client_hash_id'].nunique()}")

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Rows: 38726, distinct clients: 43


In [30]:
model_data

,content_hash_id,client_hash_id,lost_clicks,impressions,clicks,weighted_avg_position,expected_ctr,search_volume,competition_level,content_type,days_since_update_capped,word_count
0,content_b7e512995f79d5a6,client_73cda7b4e4f265ea,0.682420,1140.0,2.0,4.450877,0.002353,20,LOW,keyword article,<NA>,<NA>
1,content_905aa32a0230694e,client_73cda7b4e4f265ea,0.310596,132.0,0.0,6.363636,0.002353,90,HIGH,keyword article,<NA>,<NA>
2,content_d056587ff7faca0c,client_73cda7b4e4f265ea,-9.482190,2770.0,16.0,3.950542,0.002353,50,LOW,keyword article,<NA>,2475
3,content_2662845f598544ef,client_73cda7b4e4f265ea,-0.665874,142.0,1.0,7.929577,0.002353,30,LOW,keyword article,<NA>,<NA>
4,content_712c365258cee05c,client_73cda7b4e4f265ea,-8.769056,6048.0,23.0,4.931878,0.002353,30,LOW,keyword article,<NA>,2809
...,...,...,...,...,...,...,...,...,...,...,...,...
38721,content_31dd32e3f37395e9,client_08a6a72ff48e62c0,0.301184,128.0,0.0,7.679688,0.002353,30,LOW,keyword article,<NA>,2745
38722,content_6df919b2a32334c1,client_2094c6eb080311d5,0.178080,80.0,0.0,10.600000,0.002226,110,LOW,keyword article,<NA>,3126
38723,content_92ac4b71d16121e1,client_73cda7b4e4f265ea,0.160004,68.0,0.0,6.367647,0.002353,140,HIGH,keyword article,<NA>,2841
38724,content_aa8863866aa75488,client_73cda7b4e4f265ea,-0.603649,141.0,1.0,2.687943,0.002811,90,LOW,keyword article,<NA>,2732


In [31]:
#feature preparation
model_data['has_word_count'] = model_data['word_count'].notna().astype(int)
model_data['word_count'] = model_data['word_count'].fillna(0)

model_data['updated_before_march31'] = model_data['days_since_update_capped'].notna().astype(int)
model_data['days_since_update_capped'] = model_data['days_since_update_capped'].fillna(0)

model_data = pd.get_dummies(model_data, columns=['competition_level'],drop_first = True)
# content_type is constant ('keyword article') in this scored slice, so it isn't a usable feature here
model_data['capture_ratio'] = model_data['clicks'] / model_data['search_volume'] # baseline rule's score — lower = higher priority

honest_features = ['search_volume', 'word_count', 'has_word_count',
                    'days_since_update_capped', 'updated_before_march31'] + \
                   [c for c in model_data.columns if c.startswith('competition_level_')]
print("Features used:", honest_features)

Features used: ['search_volume', 'word_count', 'has_word_count', 'days_since_update_capped', 'updated_before_march31', 'competition_level_LOW', 'competition_level_MEDIUM']


In [33]:
def precision_at_k(true_ids_ranked, pred_ids_ranked, k=20):
  true_top_k = set(true_ids_ranked[:k])
  pred_top_k = set(pred_ids_ranked[:k])
  return len(true_top_k & pred_top_k) / k

def run_grouped_cv(data, feature_cols, k):
  gkf = GroupKFold(n_splits = k)
  groups = data['client_hash_id']
  X_all = data[feature_cols]
  y_all = data['lost_clicks']
  fold_results = []

  for fold_i, (train_idx, test_idx) in enumerate(gkf.split(X_all, y_all, groups = groups)):
    train_df = data.iloc[train_idx]
    test_df = data.iloc[test_idx].copy()

    X_train, y_train = train_df[feature_cols], train_df['lost_clicks']
    X_test, y_test = test_df[feature_cols], test_df['lost_clicks']

    #ground truth ranking for this fold's held-out rows
    true_ranked_ids = test_df.sort_values('lost_clicks', ascending = False)['content_hash_id'].tolist()

    #baseline
    baseline_ranked_ids = test_df.sort_values('capture_ratio', ascending = True)['content_hash_id'].tolist()
    baseline_p20 = precision_at_k(true_ranked_ids, baseline_ranked_ids, k = 20)

    #linear regression
    lin = LinearRegression().fit(X_train, y_train)
    test_df['pred_lr'] = lin.predict(X_test)
    lr_ranked_ids = test_df.sort_values('pred_lr', ascending = False)['content_hash_id'].tolist()
    lr_p20 = precision_at_k(true_ranked_ids, lr_ranked_ids, k= 20)
    lr_r2 = r2_score(y_test, test_df['pred_lr'])

    #Random Forest
    rf = RandomForestRegressor(n_estimators = 300, random_state = 42, n_jobs = -1).fit(X_train, y_train)
    test_df['pred_rf'] = rf.predict(X_test)
    rf_ranked_ids = test_df.sort_values('pred_rf', ascending = False)['content_hash_id'].to_list()
    rf_p20 = precision_at_k(true_ranked_ids, rf_ranked_ids, k = 20)
    rf_r2 = r2_score(y_test, test_df['pred_rf'])

    fold_results.append({
        'fold': fold_i,
        'n_test_clients': test_df['client_hash_id'].nunique(),
        'n_test_rows': len(test_df),
        'baseline_p20': baseline_p20,
        'lr_p20': lr_p20,
        'lr_r2': lr_r2,
        'rf_p20': rf_p20,
        'rf_r2': rf_r2
    })
  return pd.DataFrame(fold_results)


In [34]:
results_k5 = run_grouped_cv(model_data, honest_features, k = 5)
results_k10 = run_grouped_cv(model_data, honest_features, k = 10)

print("k5 Folds")
print(results_k5)
print()
print("k10 Folds")
print(results_k10)


k5 Folds
   fold  n_test_clients  n_test_rows  baseline_p20  lr_p20     lr_r2  rf_p20  \
0     0               1        11127           0.0     0.0  0.003324    0.00   
1     1               1         7163           0.0     0.0 -0.018591    0.00   
2     2              14         6812           0.0     0.0 -0.000649    0.00   
3     3              13         6812           0.0     0.0  0.001776    0.00   
4     4              14         6812           0.0     0.0 -0.235693    0.05   

       rf_r2  
0  -0.049760  
1  -0.512971  
2  -0.049015  
3  -1.126780  
4 -12.605995  

k10 Folds
   fold  n_test_clients  n_test_rows  baseline_p20  lr_p20     lr_r2  rf_p20  \
0     0               1        11127          0.00     0.0  0.003324     0.0   
1     1               1         7163          0.00     0.0 -0.018591     0.0   
2     2               1         4415          0.00     0.0  0.001104     0.0   
3     3               1         4118          0.00     0.0  0.003137     0.0   
4     4  

In [36]:
cols = ['baseline_p20', 'lr_p20', 'lr_r2', 'rf_p20', 'rf_r2']

summary = pd.DataFrame({
    'k=5':  results_k5[cols].mean(),
    'k=10': results_k10[cols].mean(),
})

summary

,k=5,k=10
baseline_p20,0.000000,0.005000
lr_p20,0.000000,0.000000
lr_r2,-0.049966,-0.048622
rf_p20,0.010000,0.010000
rf_r2,-2.868904,-3.279192


In [37]:
print(model_data[['lost_clicks', 'impressions']].corr(method='spearman'))

pos = model_data['lost_clicks'].clip(lower=0)
print("Share of all positive lost_clicks held by the top 20 pages:",
      pos.nlargest(20).sum() / pos.sum())

print(model_data.nlargest(20, 'lost_clicks')[['impressions', 'clicks', 'search_volume', 'lost_clicks']].describe())

print("Chance-level precision@20 per fold:", (20 / results_k5['n_test_rows']).round(4).tolist())

             lost_clicks  impressions
lost_clicks     1.000000     0.113221
impressions     0.113221     1.000000
Share of all positive lost_clicks held by the top 20 pages: 0.06238935343379827
         impressions      clicks  search_volume  lost_clicks
count      20.000000   20.000000      20.000000    20.000000
mean    94916.500000   53.600000    2762.500000   190.630469
std     43032.470774   80.680266    9278.280169    82.343814
min     46781.000000    1.000000      20.000000    97.408441
25%     63423.000000    4.500000      20.000000   137.965452
50%     77341.500000   29.000000      60.000000   157.355749
75%    126204.500000   54.250000     135.000000   235.525102
max    203497.000000  289.000000   40500.000000   378.440024
Chance-level precision@20 per fold: [0.0018, 0.0028, 0.0029, 0.0029, 0.0029]


In [38]:
from huggingface_hub import HfApi
files = HfApi().list_repo_files("FlyRank/internship-warehouse", repo_type="dataset")
[f for f in files if "daily_performance" in f and "2026-02" in f]

['fact_content_daily_performance/month=2026-02/data_0.parquet']

In [39]:
feb = con.sql(f"""
    SELECT content_hash_id,
           SUM(gsc_impressions) AS feb_impressions,
           SUM(gsc_clicks) AS feb_clicks
    FROM read_parquet('{rel}/fact_content_daily_performance/month=2026-02/data_0.parquet')
    WHERE gsc_data_available = TRUE
    GROUP BY content_hash_id
""").df()

model_data = model_data.merge(feb, on='content_hash_id', how='left')
model_data['has_feb_data'] = model_data['feb_impressions'].notna().astype(int)
model_data[['feb_impressions', 'feb_clicks']] = model_data[['feb_impressions', 'feb_clicks']].fillna(0)

print("Share of pages with February data:", model_data['has_feb_data'].mean().round(3))

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Share of pages with February data: 0.877


In [40]:
print(model_data[['lost_clicks', 'feb_impressions']].corr(method='spearman'))
print(model_data[['impressions', 'feb_impressions']].corr(method='spearman'))

                 lost_clicks  feb_impressions
lost_clicks         1.000000         0.125273
feb_impressions     0.125273         1.000000
                 impressions  feb_impressions
impressions         1.000000         0.753036
feb_impressions     0.753036         1.000000


In [41]:
feb_features = honest_features + ['feb_impressions', 'feb_clicks', 'has_feb_data']

def top_k_share(test_df, ranked_ids, k=100):
    pos = test_df['lost_clicks'].clip(lower=0)
    total = pos.sum()
    picked = pos[test_df['content_hash_id'].isin(ranked_ids[:k]).values].sum()
    return picked / total

def run_grouped_cv_feb(data, feature_cols, k):
    gkf = GroupKFold(n_splits=k)
    rows = []
    for fold_i, (tr, te) in enumerate(gkf.split(data, data['lost_clicks'], groups=data['client_hash_id'])):
        train_df, test_df = data.iloc[tr], data.iloc[te].copy()
        X_train, y_train = train_df[feature_cols], train_df['lost_clicks']
        X_test, y_test = test_df[feature_cols], test_df['lost_clicks']

        true_ids = test_df.sort_values('lost_clicks', ascending=False)['content_hash_id'].tolist()
        ranks = {
            'rule': test_df.sort_values('capture_ratio', ascending=True)['content_hash_id'].tolist(),
            'febonly': test_df.sort_values('feb_impressions', ascending=False)['content_hash_id'].tolist(),
        }
        lin = LinearRegression().fit(X_train, y_train)
        test_df['pred_lr'] = lin.predict(X_test)
        ranks['lr'] = test_df.sort_values('pred_lr', ascending=False)['content_hash_id'].tolist()

        rf = RandomForestRegressor(n_estimators=300, random_state=42, n_jobs=-1).fit(X_train, y_train)
        test_df['pred_rf'] = rf.predict(X_test)
        ranks['rf'] = test_df.sort_values('pred_rf', ascending=False)['content_hash_id'].tolist()

        row = {'fold': fold_i, 'n_test_rows': len(test_df),
               'chance_share100': min(100 / len(test_df), 1.0),
               'lr_r2': r2_score(y_test, test_df['pred_lr']),
               'rf_r2': r2_score(y_test, test_df['pred_rf'])}
        for name, ids in ranks.items():
            row[f'{name}_p20'] = precision_at_k(true_ids, ids, k=20)
            row[f'{name}_share100'] = top_k_share(test_df, ids, k=100)
        rows.append(row)
    return pd.DataFrame(rows)

res5 = run_grouped_cv_feb(model_data, feb_features, k=5)
res10 = run_grouped_cv_feb(model_data, feb_features, k=10)
print(res5.round(3)); print(); print(res10.round(3))

   fold  n_test_rows  chance_share100  lr_r2  rf_r2  rule_p20  rule_share100  \
0     0        11127            0.009  0.739  0.158       0.0          0.006   
1     1         7163            0.014  0.652  0.553       0.0          0.004   
2     2         6812            0.015  0.358  0.334       0.0          0.013   
3     3         6812            0.015  0.874  0.790       0.0          0.017   
4     4         6812            0.015  0.444  0.023       0.0          0.008   

   febonly_p20  febonly_share100  lr_p20  lr_share100  rf_p20  rf_share100  
0         0.15             0.093    0.45        0.181    0.15        0.130  
1         0.25             0.135    0.40        0.217    0.20        0.199  
2         0.10             0.177    0.35        0.219    0.30        0.203  
3         0.20             0.157    0.60        0.241    0.40        0.189  
4         0.20             0.105    0.25        0.205    0.25        0.184  

   fold  n_test_rows  chance_share100  lr_r2  rf_r2  rul

In [42]:
cols = ['rule_p20', 'febonly_p20', 'lr_p20', 'rf_p20',
        'rule_share100', 'febonly_share100', 'lr_share100', 'rf_share100',
        'chance_share100', 'lr_r2', 'rf_r2']
pd.DataFrame({'k=5': res5[cols].mean(), 'k=10': res10[cols].mean()}).round(3)

,k=5,k=10
rule_p20,0.000,0.005
febonly_p20,0.180,0.220
lr_p20,0.410,0.395
rf_p20,0.260,0.250
rule_share100,0.010,0.036
febonly_share100,0.134,0.228
lr_share100,0.213,0.276
rf_share100,0.181,0.241
chance_share100,0.013,0.039
lr_r2,0.613,0.445


In [43]:
feb_pos = con.sql(f"""
    WITH daily AS (
        SELECT content_hash_id, gsc_avg_position, gsc_impressions, gsc_clicks
        FROM read_parquet('{rel}/fact_content_daily_performance/month=2026-02/data_0.parquet')
        WHERE gsc_data_available = TRUE AND gsc_avg_position > 0 AND gsc_impressions > 0
    )
    SELECT content_hash_id,
           SUM(gsc_impressions) AS f_impr, SUM(gsc_clicks) AS f_clicks,
           SUM(gsc_avg_position * gsc_impressions) / SUM(gsc_impressions) AS f_pos
    FROM daily GROUP BY content_hash_id
""").df()

feb_pos['f_expected_ctr'] = np.select(
    [feb_pos['f_pos'] <= 3, feb_pos['f_pos'] <= 10, feb_pos['f_pos'] <= 20],
    [0.002811, 0.002353, 0.002226], default=0.000748)
feb_pos['feb_lost_clicks'] = feb_pos['f_expected_ctr'] * feb_pos['f_impr'] - feb_pos['f_clicks']

# run this merge once only
model_data = model_data.merge(feb_pos[['content_hash_id', 'feb_lost_clicks']], on='content_hash_id', how='left')
model_data['feb_lost_clicks'] = model_data['feb_lost_clicks'].fillna(0)

def eval_rank_column(data, col, n_folds):
    out = []
    for tr, te in GroupKFold(n_splits=n_folds).split(data, data['lost_clicks'], groups=data['client_hash_id']):
        t = data.iloc[te]
        true_ids = t.sort_values('lost_clicks', ascending=False)['content_hash_id'].tolist()
        ids = t.sort_values(col, ascending=False)['content_hash_id'].tolist()
        out.append({'p20': precision_at_k(true_ids, ids, 20), 'share100': top_k_share(t, ids, 100)})
    return pd.DataFrame(out).mean()

print("k=5:\n", eval_rank_column(model_data, 'feb_lost_clicks', 5))
print("k=10:\n", eval_rank_column(model_data, 'feb_lost_clicks', 10))

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

k=5:
 p20         0.430000
share100    0.223787
dtype: float64
k=10:
 p20         0.390000
share100    0.271544
dtype: float64


In [44]:
top_true = model_data.nlargest(100, 'lost_clicks')
print(top_true['has_feb_data'].mean())              # how many even had February data
print((top_true['feb_lost_clicks'] <= 0).mean())    # share with no February gap
print(top_true['feb_lost_clicks'].describe())

1.0
0.11
count    100.000000
mean      50.378051
std      102.034709
min      -35.520911
25%        6.324605
50%       26.414660
75%       44.579731
max      569.760211
Name: feb_lost_clicks, dtype: float64


In [45]:
top_true = model_data.nlargest(100, 'lost_clicks').copy()
model_data['feb_rank'] = model_data['feb_lost_clicks'].rank(ascending=False, method='min')
top_true['feb_rank'] = model_data.loc[top_true.index, 'feb_rank']
print(top_true['feb_rank'].describe())
print("Share ranked in Feb top 100:", (top_true['feb_rank'] <= 100).mean())
print("Share ranked in Feb top 1000:", (top_true['feb_rank'] <= 1000).mean())

count      100.000000
mean      5194.210000
std      11263.754192
min          1.000000
25%         39.750000
50%        146.500000
75%       1377.000000
max      38178.000000
Name: feb_rank, dtype: float64
Share ranked in Feb top 100: 0.43
Share ranked in Feb top 1000: 0.72


## 3. Train + compare vs my baseline

*Same data, same metric, same split as your Week-4 baseline. Show the table.*

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 4. Errors and interpretation

*Where is the model wrong? What does it lean on? A short error analysis beats a big metric table.*

In [46]:
from sklearn.inspection import permutation_importance

feb_features = honest_features + ['feb_impressions', 'feb_clicks', 'has_feb_data']

n = len(model_data)
pred_lr = np.zeros(n)
fold_id = np.zeros(n, dtype=int)
imp_tables = []

for i, (tr, te) in enumerate(GroupKFold(n_splits=5).split(
        model_data, model_data['lost_clicks'], groups=model_data['client_hash_id'])):
    lin = LinearRegression().fit(model_data.iloc[tr][feb_features], model_data.iloc[tr]['lost_clicks'])
    pred_lr[te] = lin.predict(model_data.iloc[te][feb_features])
    fold_id[te] = i
    pi = permutation_importance(lin, model_data.iloc[te][feb_features],
                                model_data.iloc[te]['lost_clicks'],
                                n_repeats=10, random_state=42, scoring='r2')
    imp_tables.append(pd.DataFrame({'feature': feb_features, 'importance': pi.importances_mean}))

imp = pd.concat(imp_tables).groupby('feature')['importance'].agg(['mean', 'min', 'max'])
imp.sort_values('mean', ascending=False).round(4)

,mean,min,max
feature,,,
feb_clicks,2.4227,0.7881,3.6987
feb_impressions,0.6250,0.0067,1.6207
has_feb_data,0.0023,0.0001,0.0101
search_volume,0.0005,-0.0001,0.0028
competition_level_LOW,0.0004,-0.0000,0.0013
competition_level_MEDIUM,0.0003,-0.0003,0.0014
updated_before_march31,0.0000,-0.0000,0.0001
days_since_update_capped,0.0000,-0.0000,0.0001
has_word_count,-0.0000,-0.0014,0.0010


In [47]:
oof = model_data.copy()
oof['fold'] = fold_id
oof['pred_lr'] = pred_lr
for col, name in [('lost_clicks', 'true'), ('pred_lr', 'lr'), ('feb_lost_clicks', 'persist')]:
    oof[f'{name}_rank'] = oof.groupby('fold')[col].rank(ascending=False, method='first')

show = ['fold', 'lost_clicks', 'feb_lost_clicks', 'pred_lr', 'impressions', 'feb_impressions',
        'clicks', 'feb_clicks', 'weighted_avg_position', 'has_feb_data',
        'true_rank', 'lr_rank', 'persist_rank']

false_alarms = oof[(oof['lr_rank'] <= 20) & (oof['true_rank'] > 1000)]
misses = oof[(oof['true_rank'] <= 20) & (oof['lr_rank'] > 1000)]
print("Model top-20 but truly rank >1000:", len(false_alarms))
print("Truly top-20 but model ranks >1000:", len(misses))

false_alarms.sort_values('pred_lr', ascending=False)[show].head(3)

Model top-20 but truly rank >1000: 11
Truly top-20 but model ranks >1000: 20


,fold,lost_clicks,feb_lost_clicks,pred_lr,impressions,feb_impressions,clicks,feb_clicks,weighted_avg_position,has_feb_data,true_rank,lr_rank,persist_rank
35066,0,0.711424,130.361000,119.207472,2288.0,51000.0,1.0,13.0,21.768794,1,3983.0,7.0,5.0
7402,4,-16.641776,104.048836,106.825466,91388.0,79986.0,85.0,74.0,20.481726,1,6649.0,2.0,2.0
16457,2,-10.916523,52.579760,48.911094,59109.0,55920.0,150.0,79.0,5.214925,1,6349.0,3.0,3.0


In [48]:
misses.sort_values('lost_clicks', ascending=False)[show].head(3)

,fold,lost_clicks,feb_lost_clicks,pred_lr,impressions,feb_impressions,clicks,feb_clicks,weighted_avg_position,has_feb_data,true_rank,lr_rank,persist_rank
22677,1,293.523707,-4.767706,-8.732911,143019.0,28998.0,43.0,73.0,3.166132,1,1.0,6712.0,6559.0
25893,0,165.270089,1.451851,2.053708,71513.0,1467.0,3.0,2.0,6.983444,1,4.0,2118.0,2286.0
11134,2,158.822719,-18.708399,-22.029186,89229.0,17891.0,92.0,69.0,2.410965,1,3.0,6652.0,6580.0


In [49]:
from scipy.stats import spearmanr
for f in range(5):
    t = oof[oof['fold'] == f]
    print(f, round(spearmanr(t['pred_lr'], t['feb_lost_clicks'])[0], 3))

0 0.939
1 0.886
2 0.81
3 0.848
4 0.823


In [50]:
for f in range(5):
    t = oof[oof['fold'] == f]
    top_lr = set(t.nlargest(100, 'pred_lr')['content_hash_id'])
    top_p = set(t.nlargest(100, 'feb_lost_clicks')['content_hash_id'])
    print(f, len(top_lr & top_p), "of 100 shared")

0 91 of 100 shared
1 87 of 100 shared
2 60 of 100 shared
3 86 of 100 shared
4 76 of 100 shared


In [51]:
rows = []
for f in range(5):
    t = oof[oof['fold'] == f].copy()
    t['pos_true'] = t['lost_clicks'].clip(lower=0)
    top_lr = set(t.nlargest(100, 'pred_lr')['content_hash_id'])
    top_p = set(t.nlargest(100, 'feb_lost_clicks')['content_hash_id'])
    only_lr = t[t['content_hash_id'].isin(top_lr - top_p)]
    only_p = t[t['content_hash_id'].isin(top_p - top_lr)]
    rows.append({'fold': f, 'n_only_model': len(only_lr), 'n_only_rule': len(only_p),
                 'true_lost_only_model': only_lr['pos_true'].sum(),
                 'true_lost_only_rule': only_p['pos_true'].sum()})
disagree = pd.DataFrame(rows)
disagree['model_minus_rule'] = disagree['true_lost_only_model'] - disagree['true_lost_only_rule']
disagree.round(1)

,fold,n_only_model,n_only_rule,true_lost_only_model,true_lost_only_rule,model_minus_rule
0,0,9,9,197.2,220.9,-23.7
1,1,13,13,325.0,392.7,-67.7
2,2,40,40,273.9,344.7,-70.8
3,3,14,14,32.6,196.1,-163.5
4,4,24,24,123.4,99.8,23.7


In [52]:
import sklearn
print("scikit-learn", sklearn.__version__, "| pandas", pd.__version__, "| numpy", np.__version__)  # library versions, seeds fixed at 42

def eval_rank_column_folds(data, col, n_folds):
    out = []
    for tr, te in GroupKFold(n_splits=n_folds).split(data, data['lost_clicks'], groups=data['client_hash_id']):
        t = data.iloc[te]
        true_ids = t.sort_values('lost_clicks', ascending=False)['content_hash_id'].tolist()
        ids = t.sort_values(col, ascending=False)['content_hash_id'].tolist()
        out.append({'p20': precision_at_k(true_ids, ids, 20), 'share100': top_k_share(t, ids, 100)})
    return pd.DataFrame(out)

pers5 = eval_rank_column_folds(model_data, 'feb_lost_clicks', 5)
pers10 = eval_rank_column_folds(model_data, 'feb_lost_clicks', 10)

def fmt(s, d=3):
    return f"{s.mean():.{d}f} ({s.min():.{d}f} to {s.max():.{d}f})"

rows = []
for k, res, pers in [(5, res5, pers5), (10, res10, pers10)]:
    methods = {
        'W04 rule (capture_ratio)':        (res['rule_p20'],    res['rule_share100']),
        'Feb impressions only (no model)': (res['febonly_p20'], res['febonly_share100']),
        'Carry-forward: Feb lost_clicks':  (pers['p20'],        pers['share100']),
        'Linear Regression':               (res['lr_p20'],      res['lr_share100']),
        'Random Forest':                   (res['rf_p20'],      res['rf_share100']),
    }
    for name, (p, s) in methods.items():
        rows.append({'k': k, 'method': name, 'precision@20': fmt(p), 'share of lost clicks in top 100': fmt(s)})
    rows.append({'k': k, 'method': 'Random picking (base rate)',
                 'precision@20': f"{(20 / res['n_test_rows']).mean():.4f}",
                 'share of lost clicks in top 100': fmt(res['chance_share100'])})

final_table = pd.DataFrame(rows)
final_table

r2_table = pd.DataFrame({
    'k=5':  {'Linear Regression R2': fmt(res5['lr_r2'], 2), 'Random Forest R2': fmt(res5['rf_r2'], 2)},
    'k=10': {'Linear Regression R2': fmt(res10['lr_r2'], 2), 'Random Forest R2': fmt(res10['rf_r2'], 2)},
})
r2_table


scikit-learn 1.6.1 | pandas 2.2.3 | numpy 2.1.3


,k=5,k=10
Linear Regression R2,0.61 (0.36 to 0.87),0.45 (0.07 to 0.96)
Random Forest R2,0.37 (0.02 to 0.79),0.19 (-0.75 to 0.88)


In [53]:
print(final_table.to_string(index=False))

 k                          method           precision@20 share of lost clicks in top 100
 5        W04 rule (capture_ratio) 0.000 (0.000 to 0.000)          0.010 (0.004 to 0.017)
 5 Feb impressions only (no model) 0.180 (0.100 to 0.250)          0.134 (0.093 to 0.177)
 5  Carry-forward: Feb lost_clicks 0.430 (0.250 to 0.550)          0.224 (0.182 to 0.289)
 5               Linear Regression 0.410 (0.250 to 0.600)          0.213 (0.181 to 0.241)
 5                   Random Forest 0.260 (0.150 to 0.400)          0.181 (0.130 to 0.203)
 5      Random picking (base rate)                 0.0027          0.013 (0.009 to 0.015)
10        W04 rule (capture_ratio) 0.005 (0.000 to 0.050)          0.036 (0.004 to 0.063)
10 Feb impressions only (no model) 0.220 (0.100 to 0.350)          0.228 (0.093 to 0.356)
10  Carry-forward: Feb lost_clicks 0.390 (0.100 to 0.700)          0.272 (0.090 to 0.402)
10               Linear Regression 0.395 (0.050 to 0.700)          0.276 (0.087 to 0.397)
10        

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.